File: 3462.maximum-sum-with-at-most-k-elements.cpp
#
# @lc app=leetcode id=3462 lang=cpp
#
# [3462] Maximum Sum With at Most K Elements
#
# @lc code=start
class Solution {
public:
    long long maxSum(vector<vector<int>>& grid, vector<int>& limits, int k) {
        priority_queue<int> maxHeap;
        for (int i = 0; i < grid.size(); ++i) {
            sort(grid[i].rbegin(), grid[i].rend()); // Sort each row in descending order
            for (int j = 0; j < min(limits[i], (int)grid[i].size()); ++j) {
                maxHeap.push(grid[i][j]); // Add up to limits[i] elements from each row to the heap
            }
        }
        long long sum = 0;
        while (k-- > 0 && !maxHeap.empty()) {
            sum += maxHeap.top();
            maxHeap.pop(); // Extract top k elements from heap for maximum sum
        }
        return sum;
    }
};
# @lc code=end

File: 3462.maximum-sum-with-at-most-k-elements.java
#
# @lc app=leetcode id=3462 lang=java
#
# [3462] Maximum Sum With at Most K Elements
#
# @lc code=start
class Solution {
    public long maxSum(int[][] grid, int[] limits, int k) {
        List<Integer> candidates = new ArrayList<>();
        for (int i = 0; i < grid.length; i++) {
            int[] row = grid[i];
            Arrays.sort(row); // ascending
            int cnt = Math.min(limits[i], row.length);
            for (int j = 0; j < cnt; j++) {
                candidates.add(row[row.length - 1 - j]); // pick largest cnt elements
            }
        }
        Collections.sort(candidates, Collections.reverseOrder());
        long sum = 0;
        for (int i = 0; i < Math.min(k, candidates.size()); i++) {
            sum += candidates.get(i);
        }
        return sum;
    }
}
# @lc code=end